feat: Native container serialization

ShadauxCat opened the pull request:
Support for native containers in FastBufferReader/Writer, BufferSerializer, NetworkVariable, and RPCs.

## Changelog

- Added: Added support for native containers in FastBufferReader/Writer, BufferSerializer, NetworkVariable, and RPCs.
## Testing and Documentation

- Includes unit tests.
- Includes integration tests.
- Includes documentation for previously-undocumented public API entry points.

ShadauxCat committed:
Native container serialization WIP

ShadauxCat committed:
feat: Support for native containers in FastBufferReader/Writer, BufferSerializer, NetworkVariable, and RPCs.

ShadauxCat committed:
standards + changelog

0xFA11 commented on com.unity.netcode.gameobjects/Editor/CodeGen/NetworkBehaviourILPP.cs:
```diff
@@ -1968,6 +2133,26 @@ private MethodDefinition GenerateStaticHandler(MethodDefinition methodDefinition
+                {
+                    // var list = NetworkBehaviour.__createNativeList<T>();
+
+                    // This simplifies things - easier to call __createNativeList() and have the implementation in C#
+                    // than to try to actually construct a NativeList in IL. This is also more future-proof.
+
+                    // Unlike other types, NativeList<> calls ReadValueSafeInPlace instead of ReadValueSafe.
+                    // FastBufferReader doesn't support a non-in-place deserializer for NativeList in order to
+                    // avoid users using it without realizing the allocation overhead that would cost. In-place
+                    // is more efficient when an existing value exists, and when it doesn't, it's easy to create one,
+                    // which is what we do here.
```
agreed — also thanks for the comment.

NoelStephensUnity commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@ Additional documentation and release notes are available at [Multiplayer Documen
 
 ### Added
 
+- Added support for native containers in FastBufferReader/Writer, BufferSerializer, NetworkVariable, and RPCs.
```
nit-pick:
could you add (#2375) at the end of this?

ShadauxCat committed:
Merge branch 'develop' into feat/native_container_serialization

ShadauxCat committed:
Add PR # to changelog

NoelStephensUnity reviewed (approved):
LGTM!

ShadauxCat committed:
Adding support for detecting changes in non-POD values (particularly native containers and managed types) that may be changed without setting the dirty flag.

ShadauxCat committed:
Merge branch 'develop' into feat/native_container_serialization

ShadauxCat committed:
Revert unintended prefab changes.

ShadauxCat committed:
Merge remote-tracking branch 'origin/feat/native_container_serialization' into feat/native_container_serialization

ShadauxCat committed:
standards, and fixed an outdated comment.

ShadauxCat committed:
Merge branch 'develop' into feat/native_container_serialization

# Conflicts:
#	com.unity.netcode.gameobjects/CHANGELOG.md
#	com.unity.netcode.gameobjects/Tests/Editor/Serialization/BaseFastBufferReaderWriterTest.cs
#	com.unity.netcode.gameobjects/Tests/Editor/Serialization/FastBufferReaderTests.cs
#	com.unity.netcode.gameobjects/Tests/Editor/Serialization/FastBufferWriterTests.cs
#	com.unity.netcode.gameobjects/Tests/Runtime/NetworkVariableTests.cs
#	com.unity.netcode.gameobjects/Tests/Runtime/RpcTests.cs

ShadauxCat committed:
Refactored to reduce the number of new methods that had to be added. Also I forgot to check in the merge before starting this work, so...

Merge branch 'develop' into feat/native_container_serialization

# Conflicts:
#	com.unity.netcode.gameobjects/Editor/CodeGen/NetworkBehaviourILPP.cs
#	com.unity.netcode.gameobjects/Editor/CodeGen/RuntimeAccessModifiersILPP.cs

ShadauxCat committed:
Standards

ShadauxCat committed:
Test fixes

NoelStephensUnity reviewed (approved):
Looks good to me (a second time!) 😸

ShadauxCat committed:
Merge branch 'develop' into feat/native_container_serialization

ShadauxCat committed:
Updated changelog wording a little.

ShadauxCat committed:
Merge branch 'develop' into feat/native_container_serialization
